File: src/executor/state_machine/nodes/execute.node.ts
/**
 * execute 节点
 * 
 * 执行阶段，负责：
 * 1. 根据 planned_action 执行具体动作
 * 2. 使用 Playwright 执行动作
 * 3. 捕获执行结果和截图
 * 4. 处理执行失败情况
 */
import { ExecutionState, NodeResult } from '../../types';
import { SandboxManager, SandboxContext } from '../../worker/sandbox.manager';
import { ExecutorAdapter, executorAdapter } from '../../adapters/executor.adapter';

export async function executeNode(
  state: ExecutionState,
  sandbox: SandboxContext,
  sandboxManager: SandboxManager,
  executor: ExecutorAdapter = executorAdapter
): Promise<NodeResult> {
  console.log(`[Execute] 执行动作: ${state.instanceId}`);

  const { plannedAction } = state;

  if (!plannedAction) {
    console.error('[Execute] 无动作计划');
    return {
      nodeName: 'execute',
      success: false,
      nextNode: 'end',
      state: {
        ...state,
        error: 'No planned action',
      },
    };
  }

  try {
    // 执行动作
    const result = await executor.executeAction(sandbox.page, plannedAction);

    // 更新截图
    let newScreenshot = state.lastScreenshot;
    if (result.screenshot) {
      newScreenshot = result.screenshot;
    }

    if (!result.success) {
      console.log(`[Execute] 执行失败: ${result.error}`);
      return {
        nodeName: 'execute',
        success: false,
        nextNode: 'end',
        state: {
          ...state,
          lastScreenshot: newScreenshot,
          error: result.error,
        },
      };
    }

    console.log('[Execute] 执行成功');
    return {
      nodeName: 'execute',
      success: true,
      nextNode: 'verify',
      state: {
        ...state,
        currentNode: 'executed',
        lastScreenshot: newScreenshot,
        stepCount: state.stepCount + 1,
      },
    };
  } catch (error) {
    console.error('[Execute] 执行异常:', error);
    return {
      nodeName: 'execute',
      success: false,
      nextNode: 'end',
      state: {
        ...state,
        error: String(error),
      },
    };
  }
}

export default executeNode;

File: executor-python/nova_executor/nodes/execute_node.py
"""
execute 节点 - 执行节点
======================

执行阶段，负责：
1. 根据 planned_action 执行具体动作
2. 使用 Playwright 执行动作
3. 捕获执行结果和截图
4. 处理执行失败情况

严格遵循 docs/design/02_概要设计文档_v1.0.md 的技术栈要求
- 使用 Magentic-UI (Playwright) 执行浏览器动作
"""

import logging
from typing import Dict, Any

from nova_executor.types import ExecutionState, NodeName
from nova_executor.adapters import ExecutorAdapter

logger = logging.getLogger(__name__)


async def execute_node(state: ExecutionState) -> Dict[str, Any]:
    """
    execute 节点执行函数

    调用 Executor 适配器执行动作

    Args:
        state: 当前执行状态

    Returns:
        状态更新字典
    """
    logger.info(f"[Execute] 执行动作: {state.instance_id}")

    if not state.planned_action:
        logger.error("[Execute] 无动作计划")
        return {
            "current_node": NodeName.EXECUTE,
            "hil_triggered": True,
            "error": "No planned action",
        }

    try:
        # 调用 Executor 适配器执行动作
        executor = ExecutorAdapter()
        result = await executor.execute_action(
            instance_id=state.instance_id,
            action=state.planned_action,
            previous_screenshot=state.last_screenshot,
        )

        if not result.success:
            logger.warning(f"[Execute] 执行失败: {result.error}")
            return {
                "current_node": NodeName.EXECUTE,
                "last_screenshot": result.screenshot,
                "step_count": state.step_count + 1,
                "error": result.error,
            }

        logger.info("[Execute] 执行成功")

        return {
            "current_node": NodeName.EXECUTE,
            "last_screenshot": result.screenshot,
            "step_count": state.step_count + 1,
            "hil_triggered": False,
            "error": None,
        }

    except Exception as e:
        logger.exception(f"[Execute] 执行异常: {e}")
        return {
            "current_node": NodeName.EXECUTE,
            "hil_triggered": True,
            "error": f"Execute failed: {str(e)}",
        }
